### PR TITLE
included an NDDataRef (NDData with all Mixins) class as alternative to NDDataArray.

### PR DESCRIPTION
**Pull request by MSeifert04**
The `NDDataAllMixins`-class (horrible name but I haven't figured out a better name) is an alternative to the compatibility class `NDDataArray`. It's nothing special it's an empty class that could serve for demonstration purposes in the documentation and as an useable implementation of `NDData`+Mixins for user that don't want to subclass themself.

I've changed the tests for the Mixins to use this class. But I really don't like the name.

Rationale for this class:

`NDDataArray` was kept as compatibility class during APE7 (see [Transition to astropy 1.0](https://astropy.readthedocs.org/en/v1.0.9/nddata/index.html#transition-to-astropy-1-0)). It provides all Mixins too but with a `numpy.ndarray`-like interface.

In my opinion `NDDataArray` would need a lot of work to make it recommendable. The `ndarray`-like-ness is a bit of a curse because it only keeps the `data` and `mask` (see [https://github.com/astropy/astropy/blob/master/astropy/nddata/compat.py#L237](sourcecode)) but throws away everything else. This gives unexpected results when using UFUNCS:

```
from astropy.nddata import NDDataArray, NDDataAllMixins
import numpy as np
ndd1 = NDDataAllMixins(np.ones((3,3)), 
                       meta={'cool': True}, 
                       mask=np.random.random((3,3))>0.5,
                       uncertainty=StdDevUncertainty(np.ones((3,3))),
                       unit='m')
ndd2 = NDDataArray(ndd1, copy=True)
np.sin(ndd2)
masked_array(data =
 [[0.8414709848078965 -- --]
 [0.8414709848078965 -- --]
 [0.8414709848078965 -- 0.8414709848078965]],
             mask =
 [[False  True  True]
 [False  True  True]
 [False  True False]],
       fill_value = 1e+20)
```
While it just simple doesn't work with `NDDataAllMixins`:
```
np.sin(ndd1)
AttributeError: 'NDDataAllMixins' object has no attribute 'sin'
```

So instead of telling people to subclass `NDData` and `NDArithmeticMixin` or `NDData` and `NDIOMixin` or `NDData` and `NDSlicingMixin` one could just describe `NDDataAllMixins` which has all these Mixins implemented without promoting `NDDataArray` (which also puts more restrictions on the implementation than `NDData` does).

Long story short:
The reason is because I want to rewrite parts of the `nddata`-package documentation:
- NDDataAllMixins as _the class_ the users probably want to use (Getting started and explaining the Mixins)
- NDData if the user wants just the container (no arithmetics, slicing or io-interface)
- NDDataBase if the user wants the abstract meta-container because NDData is too restricted for his use-case.
- don't mention `NDDataArray` that's just for compatibility with pre-astropy 1.0 code as far as I see it.

### CODE REVIEW

**Comment by MSeifert04** [comment]
Could someone please restart the AppVeyor build?

**Comment by pllim** [comment]
I still haven't figured out AppVeyor account stuff, but @bsipocz or @astrofrog should be able to.

**Comment by pllim** [comment]
For users who are not familiar with the inner-workings of `NDData` and compatible mixins, it is nice to explain what "all" means.

**Comment by bsipocz** [comment]
I've restarted the Appveyor build.

**Comment by astrofrog** [comment]
@pllim - can you create an account on appveyor, and I'll add you to the adminstrators

**Comment by taldcroft** [comment]
I agree that the name is "horrible".  :smile:  "All" mixins is a potentially moving target as new mixins are added.

On the bigger picture, what is the driver for adding this?  Demonstration code is appropriate for docs / tutorials but not so much in the core library.  Are we going to recommend using this instead of `NDDataArray` as the one-stop way of getting a fully-functional NDData container?  I'm worried about creating even more diversity in the NDData universe instead of working toward convergence.

**Comment by astrofrog** [comment]
@pllim - I just added all core astropy maintainers (which you are one) to appveyor (you can now add access for teams) so you should be able to log in with your github account.

**Comment by taldcroft** [comment]
Thanks @astrofrog, I've been missing that particular superpower.

**Comment by pllim** [comment]
Power

**Comment by MSeifert04** [comment]
I've edited the first post to include a bit more general information.

@pllim `NDData` is just a container. As such it doesn't support anything but there are three Mixins: `IO-interface` (io.registry), arithmetics (add, subtract, divide, multiply) and slicing. These 3 (all!) Mixins are implemented together with NDData here.

@taldcroft Yes, I wanted to recommend it as "the one-stop way of getting a fully-functional NDData container". `NDDataArray` while not officially deprecated was described as just a compatibility class to pre-astropy-1.0 and implements some stuff that was explicitly mentioned in APE7 as being problematic: very incomplete `numpy.ndarray`-like interface attempt. So I hesitate to recommend to recommend it somewhere in the documentation. It's hard to explain why the `NDArithmeticMixin` implements `add` as method instead of as `__add__` (so that `+` can be used) when:

```
from astropy.nddata import NDDataArray, NDDataAllMixins
import numpy as np
ndd1 = NDDataAllMixins(np.ones((3,3)), 
                       meta={'cool': True}, 
                       mask=np.random.random((3,3))>0.5,
                       uncertainty=StdDevUncertainty(np.ones((3,3))),
                       unit='m')
ndd2 = NDDataArray(ndd1, copy=True)
ndd2 + np.array([1,2,3])
masked_array(data =
 [[2.0 -- --]
 [2.0 -- --]
 [2.0 -- 4.0]],
             mask =
 [[False  True  True]
 [False  True  True]
 [False  True False]],
       fill_value = 1e+20)
```
But with `NDDataAllMixins` it shows the expected behaviour:
```
ndd1 + np.array([1,2,3])
# TypeError: unsupported operand type(s) for +: 'NDDataAllMixins' and 'int'

import astropy.units as u
ndd1.add(np.arange(3) * u.m) # use add-method...
#NDDataAllMixins([[ 1.,  2.,  3.],
#                 [ 1.,  2.,  3.],
#                 [ 1.,  2.,  3.]])
```
Note: Using `ndd2.add(np.arange(3) * u.m)` produces the same result as `ndd1.add(np.arange(3) * u.m)`.

**Comment by MSeifert04** [comment]
@taldcroft I also considered `NDDataSuper` or `NDSuperArray` but they are horrible too. :smile:

**Comment by taldcroft** [comment]
@MSeifert04 - OK, sounds reasonable.  I definitely like the idea of presenting a useful user-facing class in the core and the docs.  I didn't appreciate the issues with `NDDataArray`, so :+1: on this.

The name will be interesting to find since the namespace is well mined already.  I guess people would object to `NDArray`...

**Comment by pllim** [comment]
`NDDataWithRegistryArithmeticSlice` would be pretty descriptive :smile:

**Comment by bsipocz** [comment]
@pllim - ND is also an abbreviation, right? So go with the full name :smile:

**Comment by MSeifert04** [comment]
you mean something like `ArbitaryDimensionalDataArrayLikeContainerStructureWithIOCapabilitiesSimpleArithmeticsAndSlicingMixins` or abbreviated `ADDACSWIOSM`?

**Comment by pllim** [comment]
`NDDataPlus`?

**Comment by taldcroft** [comment]
`NDDataRef` ?  As in the reference implementation of a useful NDData object.

**Comment by pllim** [comment]
Or a more drastic approach is to rename `NDData` to `NDDataMinimal` (or something like that) and rename this to `NDData`... but that is probably too drastic.

**Comment by taldcroft** [comment]
@pllim - I agree with you that the following name translations would make more sense to me:

- `NDDataBase`  =>  `NDDataMeta`: I usually like have `Meta` in meta classes so you know that it is not something to directly inherit from, as opposed to a `Base` class)
- `NDData` => `NDDataBase`: the current `NDData` is really a Base class since it is not useful on its own.
- `NDDataWithUsefulMixins` => `NDData`: this is the astropy-core reference implementation being discussed here that will probably work for 90% of users and developers.

The main problem of course is that the horse is already out of the barn and changing the definition of the name `NDData` makes for serious backward compatibility problems.

**Comment by pllim** [comment]
@taldcroft , how about using `NDDataRef` with deprecation warning for 1.2, and changing the names like you said in 1.3?

**Comment by taldcroft** [comment]
Another issue to be aware of is that @perrygreenfield is currently working on ideas coming out of the PyAstro16 and LSST summit meetings that may propose back-incompatible changes in NDData.  So we probably should wait to see what Perry circulates.

**Comment by MSeifert04** [comment]
@taldcroft @perrygreenfield Are there any specs for planned changes?

But I agree if they implement backwards-incompatible changes renaming classes should come at the same time. So should I rename it to `NDDataRef`?

**Comment by pllim** [comment]
:+1: to `NDDataRef`

**Comment by pllim** [comment]
But there should be a follow-up issue about the class renaming that @taldcroft suggested so we don't forget about it.

**Comment by MSeifert04** [comment]
I've changed the name. But I was wondering we could just name it `NDD` (for n-dimensional-data). That would be consistent with most existing documentation where instances were named `ndd1`, `ndd2`, ...

It wouldn't be a nice name (could be mistaken as abbreviation for `NDData`) but I feel like `NDDataRef` also sounds like something that will be renamed soon :smile:

It's annoying that changing names is one of the most annoying backwards-incompatible changes one can make. An alternative would be to create a new subpackage (for example datacontainer) and deprecate the old-nddata altogether. But that rather strongly depends on how incompatible the changes for LSST will be.

**Comment by mwcraig** [comment]
> `NDDataPlus`?

@pllim -- this has the advantage of being extensible in the future to `NDDataPlusPlus`, `NDDataPlusPlusPlus`, etc.

**Comment by mwcraig** [comment]
@MSeifert04 -- haven't reviewed in detail, but I'm intrigued by the idea. Longer response tomorrow. I think to some extent we don't worry too much about the broader LSST discussion right now because that will take time to work out and we need some further improvement in `nddata` (the module) for 1.2, I think.

**Comment by MSeifert04** [comment]
@mwcraig I hope your joking with ``NDDataPlusPlus`` I don't like that, it should be ``NDDataSquared`` or ``NDDataCubed`` keeps the names shorter :smile: 

I really hope this class would make it into 1.2 - it makes documentation a lot easier and consistent and explaining the Mixins doesn't require phrases like: Now create your own subclass that contains the Mixins _in the right order_ and implements ``pass``.

**Comment by mwcraig** [comment]
> it should be `NDDataSquared` or `NDDataCubed` keeps the names shorter

Definitely superior!

For now, let's stick with `NDDataRef`, and I'll open an issue with @taldcroft's suggestions. I think that broader renaming will make more sense in the context of Perry's broader APE.

@MSeifert04 -- could you please rebase this one next?

**Comment by MSeifert04** [comment]
PR is rebased.

**Comment by mwcraig** [comment]
Thanks -- will merge mid-day tomorrow[1] unless @pllim or @taldcroft object (or anyone else, for that matter).

I don't see this replacing `NDDataArray` quite yet (pulling it out will require some work over in `ccdproc`), but this is a cleaner, clearer implementation of something like a functional `NDData`.

[1] On Wed, May 11

**Comment by pllim** [comment]
No problem from me. ps. I kinda like `NDDataPlusPlusPlus` but this will have to do. :wink:

**Comment by mwcraig** [comment]
Merging, thanks again!

**Comment by MSeifert04** [comment]
@mwcraig @taldcroft Don't forget to open an issue including the (serious) suggestions about renaming. Or ping me if I should open the issue.

**Comment by MSeifert04** [comment]
Thanks everyone for helping me with this PR :+1:
